models/ed/R/model2netcdf.ED2.R

```diff
@@ -971,100 +970,115 @@ read_E_files <- function(yr, yfiles, h5_files, outdir, start_date, end_date,
-          ed.dat[[length(ed.dat)+1]] <- list()    # Add space for new ed.datiable
-          ed.dat[[length(ed.dat)]][1:(t-1)] <- NA # Give NA for all previous time points
-          ed.dat[[length(ed.dat)]][t] <- ncdf4::ncvar_get(nc, allvars[j]) # Assign the value of the new ed.datiable at this time point
-          names(ed.dat)[[length(ed.dat)]] <- allvars[j] 
-          
-        }
-      }      
+    patch_df <-
+      purrr::map(avail_patch, function(.x) as.vector(ncdf4::ncvar_get(nc, .x))) %>%
+      purrr::set_names(avail_patch) %>%
+      tibble::as_tibble() %>%
+      dplyr::mutate(PATCH_ID = 1:dplyr::n())
+    
+    # join patch data with cohort data.  PACO_N is the number of cohorts (i.e.
+    # rows) in each patch, so I use that to figure out which rows are in which
+    # patch.
+    cohort_patch_id <- c()
+    for (i in seq_len(nrow(patch_df))) {
+      cohort_patch_id <- c(cohort_patch_id, rep(i, patch_df$PACO_N[i]))
     }
+    cohort_df$PATCH_ID <- cohort_patch_id
     
-  } # end ysel-loop
-
+    # Then the cohort and patch dataframes can be joined by PATCH_ID
+    dplyr::left_join(cohort_df, patch_df, by = "PATCH_ID") %>% 
+      # Dates in filename are not valid because day is 00.  Extract just year
+      # and month and use lubridate:ym() to build date
+      dplyr::mutate(date = stringr::str_match(basename(file), "(\\d{4}-\\d{2})-\\d{2}")[, 2] %>% lubridate::ym()) %>%
+      dplyr::select(date, dplyr::everything()) %>%
+      dplyr::select(-"PACO_ID")
+  }
+  
+  # Extract data from all the .h5 files
+  raw <- 
+    purrr::map(file.path(outdir, h5_files), extract_E_file) %>% 
+    dplyr::bind_rows() 
+  
+  # Unit conversions
+  out <- raw %>% 
+    dplyr::mutate(
+      # Convert units of per plant to per m^2 by multiplying by plant density
+      # (plant/m^2)
+      dplyr::across(
+        dplyr::any_of(c("BSEEDS_CO", "AGB_CO", "MMEAN_NPPDAILY_CO", "MMEAN_TRANSP_CO")),
+        function(.x) .x * .data$NPLANT
+      ),
+      # Convert to pecan standard units
+      MMEAN_NPPDAILY_CO = PEcAn.utils::ud_convert(.data$MMEAN_NPPDAILY_CO, "kg/m^2/yr", "kg/m^2/s")
+    ) %>%
+    # Weighted summary
+    # For each month (-E- file) and PFT...
+    dplyr::group_by(.data$date, .data$PFT) %>% 
+    # ... get a weighted sum across cohorts and patches
+    dplyr::summarize(
+      dplyr::across(
+        dplyr::any_of(c("BSEEDS_CO", "AGB_CO", "MMEAN_NPPDAILY_CO", "MMEAN_TRANSP_CO", "NPLANT")),
+        function(.x) sum(.x * .data$AREA)
+      ),
+      #(or a mean when it makes sense)
+      DBH_mean = mean(.data$DBH),
+      .groups = "drop"
+    )
+  
+  # Rename variables to PEcAn standard and convert to list
+  out_list <- out %>% 
+    dplyr::select(
+      #PEcAn name    #ED2 name          
+      "AGB_PFT"    = "AGB_CO",
+      "BSEEDS"     = "BSEEDS_CO",
+      "DBH"        = "DBH_mean",
+      "NPP_PFT"    = "MMEAN_NPPDAILY_CO",
+      "TRANSP_PFT" = "MMEAN_TRANSP_CO",
+      "DENS"       = "NPLANT"
+    ) %>% as.list()
```

Comment:
I think the elements of this list need to be arrays with PFT as one of the dimensions.  Need to check with `put_E_values()` to see what it's expecting.